src/lib/ffi/ffi_pkey_algs.cpp

```diff
@@ -441,6 +465,28 @@ int botan_privkey_load_ecdsa(botan_privkey_t* key,
    }
 
 /* ElGamal specific operations */
+int botan_privkey_create_elgamal(botan_privkey_t* key, botan_rng_t rng_obj, size_t pbits)
+   {
+#if defined(BOTAN_HAS_ELGAMAL)
+
+    if ((rng_obj == nullptr) || (key == nullptr))
+      return BOTAN_FFI_ERROR_NULL_POINTER;
+
+    if (pbits < 1024) {
+      return BOTAN_FFI_ERROR_BAD_PARAMETER;
+    }
+
+    return ffi_guard_thunk(BOTAN_CURRENT_FUNCTION, [=]() -> int {
+      Botan::RandomNumberGenerator& rng = safe_get(rng_obj);
+      Botan::DL_Group group(rng, Botan::DL_Group::Strong, pbits);
```

Comment:
Test for ARM fails because generation of strong primes is slow. I unfortunately don't have immediate idea on how perf could be improved (if possible at all). I'm wondering on either blocking the test or using Prime_Subgroup (q=0). I would prefer strong primes, but from the other hand elgamal is not very popular anyway. Opinions?

Reply:
After #1413 it is probably not going to be easy to speed up safe prime gen too much. If it is taking 10 minutes under qemu, it seems like even if we really work on it and cut runtime in half that still means this single test adds 5 minutes per qemu test run, which is pretty horrible in the long term.

FWIW libgcrypt's ElGamal uses smaller subgroups. (Not that libgcrypt is necessarily the best guide for the right thing to do, but it does indicate current behavior of GPG). See `cipher/elgamal.c`.

Easiest resolution I think is, if qbits == pbits-1, generate a safe prime, otherwise a prime subgroup. Have the test use a 1024/160 prime, which should be fast, and than do safe primes in your production keygen.

Reply:
Thanks. I havn't checked libgcrypt implementation before, but now I see they mention work of Weiner in paper from  eurocrypt'96. I need to take a look at it, not familiar with it yet.

I've used your suggestion about playing with qbits and pbits and now I use smaller key in test so it will be faster